docs/proposals/enable-local-source-code-support.md

```diff
@@ -0,0 +1,301 @@
+- [4] The Build controller on signalization of local source code feature, will prepend a step with the bundle image generated in [3] on the Strategy steps, as a replacement for the current image that pulls the source code from git.
+
+_Note_: To bundle the source code, an entity ( _Shipwright CLI_ ) will require to run code that will take the local source code from the user and put it inside a container image.
+
+_Note_: On runtime, when a Shipwright BuildRun is executing, that container image will be pulled and should self-extract its bundle, making the source code available ( _placed on a shared volume_ ) for other containers ( _like BuildKit, Kaniko, et al._ ) to use it.
+
+_Note_: Any workflow using Shipwright that intends to provide support for local source code will require to re-use the Shipwright/CLI implementation or implement a standalone one.
+
+### API Changes
+
+The changes in the API will provide users the following functionalities:
+
+- [1] Specify from where ( _local directory_ ) to build a container image that will bundle the source code and that will self-extract its contents at runtime. Also known as `bundle` image.
+- [2] Specify an existing container image to use, for pulling the source code.
+
+This EP is merelely concentrated on [1], but will make [2] easily to implement at a later stage.
+
+The API changes should consider the following features:
+
+- **Ability to upload/download source code**: Users should be able to specify how to build a bundle image. Users should be able to specify which bundle image to use, if it already exists.
+
+- **Ability to provide means of authentication to a registry**: Users should be able to authenticate when pulling/pushing to their container registry of choice.
+
+- **Ability to clean-up images if desired**: Users should have means to delete bundle images after the application container image is build.
+
+- **Ability to specify what to bundle**: Users should be able to specify an absolute path to their local source code directory.
+
+- **Ability to ensure reproducible Builds for bundle images**: Users should be able to verify if source code corresponds to a particular bundle image. To achieve this we can apply the pattern of reproducible builds, by pinning images timestamps to a particular time in the past. An example of the concept on reproducible builds is explained in [here](https://reproducible-builds.org/).
+
+- **Ability to specify particular directories that should not be bundle**: Users should be able to signalize which directories to not bundle, e.g. _/vendor_ . Reasonable defaults like _.github_ should be in place. Similar to the definitions under a `.gitignore` or `.dockerignore` file.
+
+### Proposal 01: API modifications
+
+> Extend `spec.source`
+
+ Introduce `spec.source.container`, which should host all the necessary metadata to signalize the usage of local source code, this should be a go `struct` which indicates what to bundle.
```

Comment:
OpenShift has allowed OCI images to be used as a type of source in its builds. In fact, OpenShift's CI relies on this capability extensively.

With this bundle approach, extracting content from an OCI image is a subset of what is needed to extract local source in a build. In fact, you allude to this capability later in the proposal.

I believe we had discussed elsewhere making git clone just another type that can be referenced in `spec.sources`. Perhaps we can go further and make local and container image sources "just another source type" as follows:

```yaml
kind: Build
spec:
  sources:
    - name: local
      type: LocalBundle
      localBundle:
        prune: true
        repository: quay.io/adambkaplan/localbuild
      destination: src/extracted
```

And similarly, we can support an OCI image as a source. On the back end for the build controller, we would reuse a lot of the same capabilities:

```yaml
spec:
  sources:
    - name: imageSource
      type: Image
      image:
        url: quay.io/adambkaplan/srcbundle:latest
        ...
      destination: src/extracted
```

Thoughts?

Reply:
In the proposal I mentioned the following

> Due to the state of the Remote Artifacts ( _exists in the API but is not yet fully functional_ ) feature and the uncertainty of the Build Inputs Overhaul proposal, this EP proposes to do the following changes in the current state of the Build API, meaning under `spec.source`.

However I think we need this in `spec.sources`.  @adambkaplan are you ok moving this one into https://github.com/shipwright-io/build/issues/765, so that there we can discuss:

- The above idea in detail.
- Also, speak on `Build Inputs Overhaul proposal`, to understand if this will take place or not.

Reply:
I closed #652 as I think it is a distraction at this point. Perhaps we revisit as we implement more features.

> However I think we need this in spec.sources. @adambkaplan are you ok moving this one into #765, so that there we can discuss:
> 
> * The above idea in detail.

Can you please clarify what you mean by "this"?

Reply:
I agree - let's discuss the additional source type in #765